File: examples/sphero-list/sphero-list.go
package main

import (
	"github.com/hybridgroup/gobot"
	"github.com/hybridgroup/gobot/api"
	"github.com/hybridgroup/gobot/platforms/sphero"
)

func main() {
	gbot := gobot.NewGobot()
	api := api.NewAPI(gbot)
	api.Port = "8080"
	api.Start()

	adaptor := sphero.NewSpheroAdaptor("Sphero", "/dev/tty.Sphero-YBW-RN-SPP")
	driver := sphero.NewSpheroDriver(adaptor, "sphero")

	r := 122
	g := 122
	b := 122

	work := func() {
		driver.SetRGB(uint8(r), uint8(g), uint8(b))
	}

	changeColor := func(color *int, delta int) string {
		newColor := *color + delta

		if newColor > 0 && newColor < 255 {
			*color = newColor
			driver.SetRGB(uint8(r), uint8(g), uint8(b))
		}

		return "ok"
	}

	robot := gobot.NewRobot("sphero-list",
		[]gobot.Connection{adaptor},
		[]gobot.Device{driver},
		work,
	)

	robot.AddCommand("increase_red", func(params map[string]interface{}) interface{} {
		return changeColor(&r, 10)
	})

	robot.AddCommand("decrease_red", func(params map[string]interface{}) interface{} {
		return changeColor(&r, -10)
	})

	robot.AddCommand("increase_green", func(params map[string]interface{}) interface{} {
		return changeColor(&g, 10)
	})

	robot.AddCommand("decrease_green", func(params map[string]interface{}) interface{} {
		return changeColor(&g, -10)
	})

	robot.AddCommand("increase_blue", func(params map[string]interface{}) interface{} {
		return changeColor(&b, 10)
	})

	robot.AddCommand("decrease_blue", func(params map[string]interface{}) interface{} {
		return changeColor(&b, -10)
	})

	gbot.AddRobot(robot)

	gbot.Start()
}

File: examples/sphero-list/sphero-list.js
var Cylon = require('cylon');

Cylon.robot({
  name: 'sphero-list',
  connection: { name: 'sphero', adaptor: 'sphero', port: '/dev/tty.Sphero-YBW-RN-SPP' },
  device: { name: 'sphero', driver: 'sphero' },
  r: 122,
  g: 122,
  b: 122,
  work: function(my) {
    my.sphero.setRGB(parseInt(this.currentColor()));
  },
  toHex: function (c) {
    var hex = c.toString(16);
    return hex.length == 1 ? "0" + hex : hex;
  },
  currentColor: function () {
    return "0x" + this.toHex(this.r) + this.toHex(this.g) + this.toHex(this.b);
  },

  changeColor: function(color, delta) {
    switch (color) {
      case "red":
        var newColor = this.r + delta;
        if (newColor>0 && newColor<255) this.r = newColor;
        break;
      case "green":
        var newColor = this.g + delta;
        if (newColor>0 && newColor<255) this.g = newColor;
        break;
      case "blue":
        var newColor = this.b + delta;
        if (newColor>0 && newColor<255) this.b = newColor;
    }

    var color = this.currentColor();
    this.sphero.setRGB(parseInt(color));
    return color;
  },
  commands: function() {
    return {
      increase_red: function() {
        this.changeColor("red", 10);
      },
      decrease_red: function(params) {
        this.changeColor("red", -10);
      },
      increase_green: function(params) {
        this.changeColor("green", 10);
      },
      decrease_green: function(params) {
        this.changeColor("green", -10);
      },
      increase_blue: function(params) {
        this.changeColor("blue", 10);
      },
      decrease_blue: function(params) {
        this.changeColor("blue", -10);
      }
    };
  }
});

Cylon.api('http',{
  host: '0.0.0.0',
  port: '8080',
  ssl:  false
});

Cylon.start();
